File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {

        unordered_map<Node*, Node*> oldToCopy;
        oldToCopy[NULL] = NULL;

        Node* cur = head;
        while(cur) {
            Node* cp = new Node(cur -> val);
            oldToCopy[cur] = cp;
            cur = cur -> next;
        }

        cur = head;
        while (cur) {
            Node* node = oldToCopy[cur];
            node -> next = oldToCopy[cur -> next];
            node -> random = oldToCopy[cur -> random];
            cur = cur -> next;
        }

        return oldToCopy[head];
        
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        mp = {}
        mp[None] = None
        
        cur = head

        while cur:
            new_node = Node(cur.val)
            mp[cur] = new_node
            cur = cur.next

        cur = head

        while cur:
            node = mp[cur]
            node.next = mp[cur.next]
            node.random = mp[cur.random]
            cur = cur.next
        
        return mp[head]

        